### DISCUSSION TITLE
Shared Bundles is not compatible with React 18's renderToPipeableStream

### DISCUSSION THREAD

**Discussion by liamqma** [parcel-bundler/parcel#9387] [Q&A] [answered]
I've setup a minimal example to demonstrate the problem I face:
https://github.com/liamqma/reproduce-bug-parcel-react-18

Let's say that I have two entries:
- foo.html
- bar.html

They share some common modules (e.g. React), so the generated files are
- foo.84654a46.js (shared-bundle)
- bar.67b35129.js
- bar.html
- foo.72b22426.js
- foo.html

foo.html looks like
```
<!DOCTYPE html>
<html>
  <head>
    <script type="module" src="/foo.84654a46.js"></script>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/foo.72b22426.js"></script>
  </body>
</html>
```
`foo.72b22426.js` relies on `foo.84654a46.js`. We are all good as long as the loading order is correct and deterministic.

However, with React 18's renderToPipeableStream, <script async> is recommended to use. Please see [this post](https://github.com/reactwg/react-18/discussions/114) for more details.

So the SSR'ed HTML becomes
```
<!DOCTYPE html>
<html>
  <body>
    ...
    <script type="module" src="/foo.84654a46.js" async></script>
    <script type="module" src="/foo.72b22426.js" async></script>
  </body>
</html>
```
Because of the nature of async, we cannot guarantee the shared-bundle is loaded/executed before. You can checkout [the repo](https://github.com/liamqma/reproduce-bug-parcel-react-18), where I intentionally make shared-bundle 2 seconds slower to download.

I wonder how to make shared bundle play nicely with `renderToPipeableStream`?

One way I can think of is that shared bundle is loaded by JS instead of being placed on generated HTML.

Thank you for help!

**Comment by mischnic** [comment, marked as answer]
There was some work for this here: https://github.com/parcel-bundler/parcel/pull/9227
But I don't know any details on how to use that

**Reply by liamqma** [reply]
Thanks @mischnic !
